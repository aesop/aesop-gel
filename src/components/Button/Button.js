import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './Button.module.css';

const Button = ({
  children,
  className,
  dataTestRef,
  hasLightText,
  id,
  isAlternate,
  isEnabled,
  isInline,
  onClick,
  tabIndex,
  title,
  type,
}) => {
  const classSet = cx(
    styles.base,
    styles.action,
    { [styles.alternate]: isAlternate },
    { [styles.blockStyle]: !isInline },
    { [styles.disabled]: !isEnabled },
    { [styles.external]: external },
    { [styles.lightText]: hasLightText },
    { [styles.inlineStyle]: isInline },
    className,
  );

  return (
    <button
      className={classSet}
      data-test-ref={dataTestRef}
      disabled={!isEnabled}
      id={id}
      onClick={onClick}
      tabIndex={tabIndex}
      title={title}
      type={type}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.any.isRequired,
  className: PropTypes.string,
  dataTestRef: PropTypes.string,
  hasLightText: PropTypes.bool,
  id: PropTypes.string,
  isAlternate: PropTypes.bool,
  isEnabled: PropTypes.bool,
  isInline: PropTypes.bool,
  onClick: PropTypes.func,
  tabIndex: PropTypes.number,
  title: PropTypes.string.isRequired,
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
};

Button.defaultProps = {
  children: undefined,
  className: undefined,
  dataTestRef: undefined,
  hasLightText: undefined,
  id: undefined,
  isAlternate: false,
  isEnabled: true,
  isInline: false,
  onClick: undefined,
  tabIndex: undefined,
  title: undefined,
  type: 'button',
};

export default Button;
