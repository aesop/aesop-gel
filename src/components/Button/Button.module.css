@import 'styles/variables/index';

.base {
  position: relative;
  appearance: none;
  background-color: transparent;
  cursor: pointer;
  font-family: var(--font-family-sans-serif-medium);

  &,
  &:link,
  &:visited {
    text-decoration: none;
  }
}

.blockStyle {
  display: inline-block;
  min-width: 210px;
  max-width: 400px;
  padding: 23px;
  border: 1px solid var(--color-grey-rgba-20);
  transition: background-color 400ms;

  &.hasIcon {
    padding-right: 55px;

    &:hover,
    &:focus {
      svg {
        path {
          fill: var(--color-white);
        }
      }
    }
  }

  .icon {
    position: absolute;
    right: 23px;
  }

  &.internal {
    .icon {
      top: 24px;
    }
  }

  &.external {
    .icon {
      top: 26px;
    }
  }

  &:hover,
  &:focus {
    background-color: var(--color-body-copy);
    color: var(--color-white);
  }
}

.disabled {
  border-color: var(--color-white-disabled);
  color: var(--color-white-disabled);
  cursor: default;

  &:hover,
  &:focus {
    background-color: transparent;
    color: var(--color-white-disabled);
  }
}

.action {
  padding-right: 23px;
}

.inlineStyle {
  display: inline;
  padding: 0;
  border: 0;
  font-size: rem(16px);

  &.hasIcon {
    padding-right: 8px;

    .icon {
      display: inline-block;
    }

    &.internal {
      .icon {
        transform: translate(8px, 0);
      }
    }

    &.external {
      .icon {
        transform: translate(8px, 3px);
      }
    }
  }

  svg {
    path {
      fill: var(--color-body-copy);
    }
  }

  &:hover,
  &:focus {
    color: var(--color-body-copy);

    svg {
      path {
        fill: var(--color-body-copy);
      }
    }
  }
}

.internal,
.external {
  .icon {
    display: inline-block;
  }
}

.icon {
  display: inline;

  svg {
    width: 1rem;
    max-height: 1rem;

    path {
      transition: fill 150ms var(--easing-ease-in-out-cubic);
    }
  }
}

.alternate {
  background-color: var(--color-body-copy);

  &,
  &:link,
  &:visited {
    color: var(--color-white);
  }

  &:hover,
  &:focus {
    background-color: var(--color-body-copy);

    &.disabled {
      color: var(--color-white-disabled);
    }
  }

  svg {
    path {
      fill: var(--color-white);
    }
  }

  &.disabled {
    border-color: var(--color-grey-50);
    background-color: var(--color-grey-50);
    color: var(--color-white-disabled);
  }
}

.lightText {
  color: var(--color-white);

  &:focus,
  &:hover {
    color: var(--color-grey-65);
  }
}
