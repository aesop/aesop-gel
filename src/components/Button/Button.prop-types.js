import PropTypes from 'prop-types';
import { getPropTypeOptionValues } from '~/utils/propTypes';

export const BUTTON_TYPE_PROP_TYPE_OPTIONS = {
  BUTTON: 'button',
  RESET: 'reset',
  SUBMIT: 'submit',
};

export const LINK_PROP_TYPE_SHAPE = PropTypes.exact({
  href: PropTypes.string,
  to: PropTypes.string,
});

export const PROP_TYPES = {
  alternate: PropTypes.bool,
  children: PropTypes.any.isRequired,
  className: PropTypes.string,
  dataTestRef: PropTypes.string,
  disabled: PropTypes.bool,
  handleClick: PropTypes.func,
  hasLightText: PropTypes.bool,
  inline: PropTypes.bool,
  id: PropTypes.string,
  tabIndex: PropTypes.number,
  title: PropTypes.string.isRequired,
  type: PropTypes.oneOf(getPropTypeOptionValues(BUTTON_TYPE_PROP_TYPE_OPTIONS)),
};

export default PROP_TYPES;
