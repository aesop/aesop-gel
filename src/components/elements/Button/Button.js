import React from 'react';
import cx from 'classnames';
import includes from 'lodash/includes';
import { HTML } from '~/constants';
import Icon from '~/components/elements/Icon/Icon';
import PROP_TYPES from './Button.prop-types';
import styles from './Button.module.css';

const Button = ({
  alternate = false,
  children,
  className,
  dataTestRef,
  disabled = false,
  hasLightText,
  href,
  icon = false,
  id,
  inline = false,
  onClick,
  tabIndex,
  target = '_self',
  title,
  to,
  type = 'button',
}) => {
  const internal = to;
  const external = href;

  if (onClick && icon) {
    console.error('<Button /> Event Buttons cannot have an icon');
  }

  if ((to || external) && disabled) {
    console.error('<Button /> Do not disable anchors');
  }

  if (inline && alternate) {
    console.error('<Button /> No such support for alternate inline links');
  }

  const classSet = cx(
    styles.base,
    { [styles.alternate]: alternate },
    { [styles.block]: !inline },
    { [styles.disabled]: disabled },
    { [styles.external]: external },
    { [styles.hasLightText]: hasLightText },
    { [styles.hasIcon]: icon },
    { [styles.inline]: inline },
    { [styles.internal]: internal },
    className,
  );

  const commonProps = {
    title,
    id,
    tabIndex,
  };

  if (onClick || type !== 'button') {
    return (
      <button
        className={cx(classSet, { [styles.action]: onClick || type })}
        data-test-ref={dataTestRef}
        disabled={disabled}
        onClick={onClick}
        type={type}
        {...commonProps}
      >
        {children}
      </button>
    );
  } else if (internal || external) {
    let iconName = '';
    let Element = '';
    const props = {};

    /**
     * @TODO Pick routing tool
     */
    if (internal) {
      // Element = Link;
      // iconName = 'rightArrow';
      // props.to = to;

      if (!includes(href, 'http://') && !includes(href, 'https://')) {
        console.error(
          '<Button /> Make sure the value of href contains http:// or https://',
        );
      }
      Element = HTML.A;
      iconName = 'rightUpArrow';
      props.href = href;
    } else if (external) {
      if (!includes(href, 'http://') && !includes(href, 'https://')) {
        console.error(
          '<Button /> Make sure the value of href contains http:// or https://',
        );
      }
      Element = HTML.A;
      iconName = 'rightUpArrow';
      props.href = href;
    }

    return (
      <Element
        className={cx(classSet)}
        data-test-ref={dataTestRef}
        target={target}
        {...commonProps}
        {...props}
      >
        {children}
        {icon && (
          <i aria-hidden="true" className={styles.icon}>
            <Icon name={iconName} />
          </i>
        )}
      </Element>
    );
  } else {
    console.error(
      '<Button /> Make sure you pass a prop to make this work, href, to or onClick',
    );

    return null;
  }
};

Button.propTypes = PROP_TYPES;

export default Button;
