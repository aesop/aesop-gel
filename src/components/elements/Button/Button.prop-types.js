import PropTypes from 'prop-types';
import { getPropTypeOptionValues } from '~/utils/propTypes';

export const BUTTON_TYPE_PROP_TYPE_OPTIONS = {
  BUTTON: 'button',
  RESET: 'reset',
  SUBMIT: 'submit',
};

export const LINK_PROP_TYPE_SHAPE = PropTypes.exact({
  href: PropTypes.string,
  to: PropTypes.string,
});

export const PROP_TYPES = {
  alternate: PropTypes.bool,
  children: PropTypes.any.isRequired,
  className: PropTypes.string,
  dataTestRef: PropTypes.string,
  disabled: PropTypes.bool,
  hasLightText: PropTypes.bool,
  href: PropTypes.string,
  icon: PropTypes.bool,
  inline: PropTypes.bool,
  id: PropTypes.string,
  onClick: PropTypes.func,
  tabIndex: PropTypes.number,
  target: PropTypes.string,
  title: PropTypes.string.isRequired,
  to: PropTypes.string,
  type: PropTypes.oneOf(getPropTypeOptionValues(BUTTON_TYPE_PROP_TYPE_OPTIONS)),
};

export default PROP_TYPES;
