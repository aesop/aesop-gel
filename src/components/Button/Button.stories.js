import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import { text, boolean, select } from '@storybook/addon-knobs';
import 'normalize.css';
import '~/styles/base.module.css';
import '~.storybook/storybook.module.css';
import { getPropTypeOptionValues } from '~/utils/propTypes';
import Button from './Button';
import { BUTTON_TYPE_PROP_TYPE_OPTIONS } from './Button.prop-types';

storiesOf('Button', module).add('Base component', () => (
  <Button
    className={text('className', '')}
    dataTestRef={text('dataTestRef', 'dataTestRef')}
    handleClick={action('clicked')}
    hasLightText={boolean('hasLightText', false)}
    id={text('id', '')}
    isAlternate={boolean('isAlternate', false)}
    isEnabled={boolean('isEnabled', true)}
    isInline={boolean('isInline', false)}
    tabIndex={text('tabIndex', '-1')}
    title={text('title', 'AÄ“sop')}
    type={select(
      'type',
      getPropTypeOptionValues(BUTTON_TYPE_PROP_TYPE_OPTIONS),
      'button',
    )}
  >
    {text('children', 'Button')}
  </Button>
));
