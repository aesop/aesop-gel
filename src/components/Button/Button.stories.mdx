import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import { action } from '@storybook/addon-actions';
import * as knobs from '@storybook/addon-knobs';
import { Button } from './Button';

<Meta title="Components/Button" component={Button} />

# Button

<Canvas>
  <Story name="Base component">
    <Button
      isAlternate={knobs.boolean('isAlternate', false, 'Presentation')}
      isEnabled={knobs.boolean('isEnabled', true, 'Presentation')}
      isInline={knobs.boolean('isInline', false, 'Presentation')}
      onClick={action('clicked')}
      theme={knobs.select('theme', ['dark', 'light'], 'dark', 'Presentation')}
      title={knobs.text('title', 'AÄ“sop', 'Copy')}
    >
      {knobs.text('children', 'Button')}
    </Button>
  </Story>
</Canvas>

## API

<ArgsTable of={Button} />
