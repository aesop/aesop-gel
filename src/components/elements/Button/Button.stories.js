import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import { text, boolean } from '@storybook/addon-knobs';
import 'normalize.css';
import '~/styles/base.module.css';
import '~/styles/storybook.module.css';
import Button from './Button';

const base = storiesOf('Elements.Button', module);

base.add('Base component', () => (
  <Button
    alternate={boolean('alternate', false)}
    href={text('href', '')}
    icon={boolean('icon', false)}
    inline={boolean('inline', false)}
    target={text('target', '_self')}
    title={text('title', 'Aēsop')}
    to={text('to', 'about')}
  >
    {text('children', 'Button')}
  </Button>
));

const variations = storiesOf('Elements.Button.Variations', module);

variations.add(
  'Disabled',
  () => (
    <Button disabled={true} onClick={action('clicked')} title="Aēsop">
      Disabled Button
    </Button>
  ),
  {
    info: {
      text: `The \`disable\` prop can only be applied to buttons with a \`onClick\` prop.`,
    },
  },
);

variations.add('Alternate', () => (
  <Button alternate={true} onClick={action('clicked')} title="Aēsop">
    Alternate Button
  </Button>
));

variations.add(
  'Disabled alternate',
  () => (
    <Button
      alternate={true}
      disabled={true}
      onClick={action('clicked')}
      title="Aēsop"
    >
      Disabled Alternate Button
    </Button>
  ),
  {
    info: {
      text: `The \`disable\` prop can only be applied to buttons with a \`onClick\` prop.`,
    },
  },
);

variations.add('Internal link', () => (
  <Button title="Aēsop" to="/about">
    Internal Link
  </Button>
));

variations.add('External link', () => (
  <Button href="http://aesop.com" title="Aēsop">
    External Link
  </Button>
));

variations.add('Inline button', () => (
  <Button inline={true} onClick={action('clicked')} title="Aēsop">
    Inline Button
  </Button>
));

variations.add('Inline internal link with icon', () => (
  <Button icon={true} inline={true} title="Aēsop" to="/about">
    Inline Internal Link with Icon
  </Button>
));

variations.add('Inline external link with icon', () => (
  <Button href="http://aesop.com" icon={true} inline={true} title="Aēsop">
    Inline External Link with Icon
  </Button>
));

variations.add('Internal link with icon', () => (
  <Button icon={true} title="Aēsop" to="/about">
    Internal Link with Icon
  </Button>
));

variations.add('External link with icon', () => (
  <Button href="http://aesop.com" icon={true} title="Aēsop">
    External Link with Icon
  </Button>
));

variations.add('Alternate internal link with icon', () => (
  <Button alternate={true} icon={true} title="Aēsop" to="/about">
    Alternate internal link with icon
  </Button>
));

variations.add('Alternate external link with icon', () => (
  <Button alternate={true} href="http://aesop.com" icon={true} title="Aēsop">
    Alternate external link with icon
  </Button>
));
