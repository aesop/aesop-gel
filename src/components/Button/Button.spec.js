import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Button from './Button';

configure({ adapter: new Adapter() });

const mockFn = jest.fn();

describe('<Button />', () => {
  it('should be defined', () => {
    expect(Button).toBeDefined();
  });

  it('renders base component correctly with `handleClick` prop', () => {
    const tree = renderer
      .create(
        <Button className="aesop" handleClick={mockFn} title="Aēsop">
          Button
        </Button>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders base component correct with `type` prop', () => {
    const tree = renderer
      .create(
        <Button className="aesop" title="Aēsop" type="submit">
          Submit
        </Button>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders base component correctly with `href` prop', () => {
    const tree = renderer
      .create(
        <Button href="https://aesop.com" title="Aēsop">
          Button
        </Button>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders base component correctly with `to` prop', () => {
    const tree = renderer
      .create(
        <Button title="Aēsop" to="about">
          Button
        </Button>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders component with icon correctly', () => {
    const tree = renderer
      .create(
        <Button title="Aēsop" to="about">
          Button
        </Button>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders disabled variation correctly', () => {
    const tree = renderer
      .create(
        <Button handleClick={mockFn} isEnabled={false} title="Aēsop">
          Disabled Button
        </Button>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders alternate variation correctly', () => {
    const tree = renderer
      .create(
        <Button handleClick={mockFn} isAlternate={true} title="Aēsop">
          Alternate Button
        </Button>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call mock function when button is clicked', () => {
    shallow(
      <Button handleClick={mockFn} title="Aēsop">
        Aēsop
      </Button>,
    ).simulate('click');

    expect(mockFn).toHaveBeenCalled();
  });
});

describe('<Button /> error handling.', () => {
  it('should return `null` if `handleClick` and `icon` are passed as props', () => {
    const component = shallow(
      <Button handleClick={mockFn} icon={true} title="Aēsop">
        Aēsop
      </Button>,
    );
    expect(component.type()).toEqual(null);
  });

  it('should return `null` if `to` or `external` are passed with `disabled` as props', () => {
    const component = shallow(
      <Button isEnabled={false} title="Aēsop" to="about">
        Aēsop
      </Button>,
    );
    expect(component.type()).toEqual(null);
  });

  it('should return `null` if `inline` and `alternate` are passed as props', () => {
    const component = shallow(
      <Button isAlternate={true} isInline={true} title="Aēsop" to="about">
        Aēsop
      </Button>,
    );
    expect(component.type()).toEqual(null);
  });

  it('should return `null` if a provided `href` prop is invalid', () => {
    const component = shallow(
      <Button href="aesop" title="Aēsop">
        Aēsop
      </Button>,
    );
    expect(component.type()).toEqual(null);
  });

  it('should return `null` if at least `to`, `href` or `handleClick` are not provided as a prop', () => {
    const component = shallow(<Button title="Aēsop">Aēsop</Button>);
    expect(component.type()).toEqual(null);
  });
});
