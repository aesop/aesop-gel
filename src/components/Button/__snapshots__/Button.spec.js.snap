// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders base component correctly with \`onClick\` prop 1`] = `
<button
  className="base blockStyle external aesop"
  disabled={false}
  onClick={[MockFunction]}
  title="AÄ“sop"
  type="button"
>
  Button
</button>
`;
